tsh.1
.\" Process this file with
.\" groff -man -Tascii foo.1
.\"
.TH TSH 1 "OCT 2014" Linux "User Manuals"
.SH NAME
tsh \- Tiny Shell for Operating Systems project
.SH SYNOPSIS
.B ./tsh
.SH DESCRIPTION
.B tsh 
is implemented to try to behave like a normal shell, 
for example, bash. In
.B tsh, we implemented both the basic points and the extra points.
We will try to give the details below.

.IP "Basic Points"

.BR "jobs control"
function is provided in
.B tsh
, such as stopping a foreground process, terminating
a running process, continuing a stopped process, and fetching
a background to foreground. After
.B fork(2)
, we use
.B setpgid(2)
and
.B tcsetpgrp(3)
to setup a process group and foreground process group. We forward
signals like
.B SIGINT
and 
.B SIGTSTP
to the foreground process group. We use
.B sigprocmask(2)
to ignore
.B SIGCHLD
when beginning to process the input commands, and unblock the signal
before waiting for the commands input. When
.B SIGCHLD
is blocked,
.B waitForCmd()
in
.IR runtime.c
will take over the task to reap the child processes. And this is the most
elegant approach to avoid the race condition between waiting for foreground
processes and reaping background processes.

.BR "builtin commands"
can be executed in
.B tsh
because we maintain a builtin commands array in the file 
.IR builtin_cmd.{c|h} .
With this design, we can add new builtin commands easily by
just adding new array elements to the builtin commands array.
And we can also have a common handle process for each builtin
commands.

.IP "Extra Points"

.BR alias/unalias
can help reduce the length of command line we need to
type in the terminal. Before the input commands are 
executed, we preprocess the commands by expanding
them using the alias pair configured by user. By
the way, we not only expand the first argv(argv[0])
but also try to expand the each field in argv.

.BR "I/O redirect"
is implemented in 
.BR tsh
, we could read from files as read from
.B stdin
and write files as write
.B stdout
because this amazing functionality. In
.B tsh
, we implement this function by using a struct named
.B io_config
which specifys the file descriptors to be used as
.B stdin
and 
.B stdout. In fored child processes, we use
.B dup2(2)
to copy the target fds to 
.B stdin(fd=0)
and 
.B stdout(fd=1) before
.B execv(2)
is called.

.BR pipe
is also implemented in
.B tsh. As mentioned in
.B I/O redirect
, 
..B struct io_config
specifies the file descriptors will be used as
.B stdin
and
.B stdout. The file desciptors which used to connect two processes is
generated by the system call
.B pipe(2)
, and pipe[0] will be used as
.B stdin
while pipe[1] will be used as
.B stdout.

.\"SH OPTIONS
.SH BUILTIN COMMANDS
.IP "cd [DIRECTORY]"
Change the current working directory specified as
.B DIRECTORY.
If no argument is provided by user, then this command
will use the home directory of current user by default.
.IP "echo [ARGUMENTS] ..."
This command will print all the provided arguments with
space as delimiter, and finnaly print a new line.
.IP "fg JOB-NUMBER"
.B fg
command take a positive integer argument as job number,
and fetch that background job with the specified job 
number to foreground.
.IP "bg JOB-NUMBER"
.B bg
command take a positive integer argument as job number,
and let that background job with the specified job number
continue running at background status.
.IP jobs
Print all the jobs with different status, such as
Done, Stopped, Running.
.IP "alias [KEY=VALUE]"
Print all the configured alias pari in dictionary order
if no argument is provided. And add the provided key
and value pair to the alias list when provided.

.IP "unalias KEY"
Remove the corresponding key and value pair in the alias list
specified by key.

.\".SH FILES
.\".I /etc/foo.conf
.\".RS
.\"The system wide configuration file. See
.\".BR foo (5)
.\"for further details.
.\".RE
.\".I ~/.foorc
.\".RS
.\"Per user configuration file. See
.\".BR foo (5)
.\"for further details.
.SH ENVIRONMENT
.IP PATH
This environment variable is used by
.IR tsh
to retrieve the external executable file search path.
.\".SH DIAGNOSTICS
.\"The following diagnostics may be issued on stderr:
.\" 
.\"Bad magic number.
.\".RS
.\"The input file does not look like an archive file.
.\".RE
.\"Old style baz segments.
.\".RS
.\".B foo
.\"can only handle new style baz segments. COBOL
.\"object libraries are not supported in this version.
.SH TO BE DONE
.B tsh
doesn't support
.B semicolon(;)
in command line.
.SH AUTHOR
Yang Yang <yangyang2016@u.northwestern.edu>
.\"SH "SEE ALSO"
.\"BR bar (1),
.\"BR foo (5),
.\"BR xyzzy (1)


skeleton/tsh.c
/***************************************************************************
 *  Title: MySimpleShell 
 * -------------------------------------------------------------------------
 *    Purpose: A simple shell implementation 
 *    Author: Stefan Birrer
 *    Version: $Revision: 1.1 $
 *    Last Modification: $Date: 2005/10/13 05:24:59 $
 *    File: $RCSfile: tsh.c,v $
 *    Copyright: (C) 2002 by Stefan Birrer
 ***************************************************************************/
/***************************************************************************
 *  ChangeLog:
 * -------------------------------------------------------------------------
 *    $Log: tsh.c,v $
 *    Revision 1.1  2005/10/13 05:24:59  sbirrer
 *    - added the skeleton files
 *
 *    Revision 1.4  2002/10/24 21:32:47  sempi
 *    final release
 *
 *    Revision 1.3  2002/10/23 21:54:27  sempi
 *    beta release
 *
 *    Revision 1.2  2002/10/15 20:37:26  sempi
 *    Comments updated
 *
 *    Revision 1.1  2002/10/15 20:20:56  sempi
 *    Milestone 1
 *
 ***************************************************************************/
#define __MYSS_IMPL__

/************System include***********************************************/
#include <stdlib.h>
#include <signal.h>
#include <string.h>
#include <stdio.h>

/************Private include**********************************************/
#include "tsh.h"
#include "io.h"
#include "interpreter.h"
#include "runtime.h"
#include "sig_handler.h"
#include "alias.h"

/************Defines and Typedefs*****************************************/
/*  #defines and typedefs should have their names in all caps.
 *  Global variables begin with g. Global constants with k. Local
 *  variables should be in all lower case. When initializing
 *  structures and arrays, line everything up in neat columns.
 */

#define BUFSIZE 80

/************Global Variables*********************************************/

/************Function Prototypes******************************************/
/* handles SIGINT and SIGSTOP signals */	
//static void sig(int);

/************External Declaration*****************************************/

/**************Implementation***********************************************/

int main (int argc, char *argv[])
{
  /* Initialize command buffer */
  char* cmdLine = (char*)malloc(sizeof(char*)*BUFSIZE);
  

  /* shell initialization */
  if (signal(SIGINT, int_handler) == SIG_ERR) PrintPError("SIGINT");
  if (signal(SIGTSTP, stp_handler) == SIG_ERR) PrintPError("SIGTSTP");
  if (signal(SIGTTIN, SIG_IGN) == SIG_ERR) PrintPError("SIGTSTP");
  if (signal(SIGTTOU, SIG_IGN) == SIG_ERR) PrintPError("SIGTSTP");
  if (signal(SIGCHLD, chld_handler) == SIG_ERR) PrintPError("SIGCHLD");
  //
  //tcsetpgrp (STDIN_FILENO, getpgrp());
  
  init_block_set();

  init_job_list();
  init_alias_list();

  while (!forceExit) /* repeat forever */
  {

    /* checks the status of background jobs */
	  // if there is any process recycled by shell
	  // then print out

	  //TODO unblock the signals
	  unblock_signals();
	  chld_handler(0);

	  /* print prompt */
	  //printf("%s@localhost %s $ ", getLogin(), getCurrentWorkingDir());

    /* read command line */
	tcsetpgrp(STDIN_FILENO, getpgrp());
    getCommandLine(&cmdLine, BUFSIZE);



    if(strcmp(cmdLine, "exit") == 0)
    {
      forceExit=TRUE;
      continue;
    }

    //TODO block the signals
    block_signals();


    CheckJobs();

    /* interpret command and line
     * includes executing of commands */
    Interpret(cmdLine);

  }
  destroy_job_list();
  destroy_alias_list();

  /* shell termination */
  free(cmdLine);
  return 0;
} /* end main */
